dot/node.go

```diff
@@ -41,9 +47,86 @@ type Node struct {
 	metricsServer *metrics.Server
 }
 
+//go:generate mockgen -source=node.go -destination=mock_node_test.go -package=$GOPACKAGE
+
+type nodeBuilderIface interface {
+	nodeInitialised(string) bool
+	initNode(config *Config) error
+	createStateService(config *Config) (*state.Service, error)
+	createNetworkService(cfg *Config, stateSrvc *state.Service, telemetryMailer telemetry.Client) (*network.Service,
+		error)
+	createRuntimeStorage(st *state.Service) (*runtime.NodeStorage, error)
+	loadRuntime(cfg *Config, ns *runtime.NodeStorage, stateSrvc *state.Service, ks *keystore.GlobalKeystore,
+		net *network.Service) error
+	createBlockVerifier(st *state.Service) (*babe.VerificationManager, error)
+	createDigestHandler(lvl log.Level, st *state.Service) (*digest.Handler, error)
+	createCoreService(cfg *Config, ks *keystore.GlobalKeystore, st *state.Service, net *network.Service,
+		dh *digest.Handler) (*core.Service, error)
+	createGRANDPAService(cfg *Config, st *state.Service, dh *digest.Handler, ks keystore.Keystore,
+		net *network.Service, telemetryMailer telemetry.Client) (*grandpa.Service, error)
+	newSyncService(cfg *Config, st *state.Service, fg gsync.FinalityGadget, verifier *babe.VerificationManager,
+		cs *core.Service, net *network.Service, telemetryMailer telemetry.Client) (*gsync.Service, error)
+	createBABEService(cfg *Config, st *state.Service, ks keystore.Keystore, cs *core.Service,
+		telemetryMailer telemetry.Client) (*babe.Service, error)
+	createSystemService(cfg *types.SystemInfo, stateSrvc *state.Service) (*system.Service, error)
+	createRPCService(params rpcServiceSettings) (*rpc.HTTPServer, error)
+}
+
+type nodeBuilder struct{}
+
+// NodeInitialized returns true if, within the configured data directory for the
+// node, the state database has been created and the genesis data has been loaded
+func NodeInitialized(basepath string) bool {
+	nodeInstance := nodeBuilder{}
+	return nodeInstance.nodeInitialised(basepath)
+}
+
+func (nodeBuilder) nodeInitialised(basepath string) bool {
+	// check if key registry exists
+	registry := path.Join(basepath, utils.DefaultDatabaseDir, "KEYREGISTRY")
+
+	_, err := os.Stat(registry)
+	if os.IsNotExist(err) {
+		logger.Debug("node has not been initialised from base path " + basepath +
+			": failed to locate KEYREGISTRY file in data directory")
+
+		return false
+	}
+
+	// initialise database using data directory
+	db, err := utils.SetupDatabase(basepath, false)
+	if err != nil {
+		logger.Debugf("failed to create database from base path %s: %s", basepath, err)
+		return false
+	}
+
+	defer func() {
+		// close database
+		err = db.Close()
+		if err != nil {
+			logger.Errorf("failed to close database: %s", err)
+		}
+	}()
+
+	// load genesis data from initialised node database
+	_, err = state.NewBaseState(db).LoadGenesisData()
+	if err != nil {
+		logger.Errorf("node has not been initialised from base path %s: %s", basepath, err)
+		return false
```

Comment:
Ideally return an error

```go
		return fmt.Errorf("cannot load genesis data: %w", err)
```

The base path and additional error wrapping can be added by the caller of this function, it shouldn't belong here.

Reply:
Updated.